schema/policy.go

```diff
@@ -137,36 +137,64 @@ type resourceFilter struct {
 	Path             *regexp.Regexp
 }
 
+//ScopingType describes how the authorization is scoped
+type ScopingType string
+
+const (
+	ScopedToTenant ScopingType = "to_tenant"
+	ScopedToDomain ScopingType = "to_domain"
+)
+
 //Authorization interface
 type Authorization interface {
 	TenantID() string
 	TenantName() string
+	DomainID() string
+	DomainName() string
 	AuthToken() string
 	Roles() []*Role
 	Catalog() []*Catalog
+	ScopingType() ScopingType
 }
 
 //BaseAuthorization is base struct for Authorization
 type BaseAuthorization struct {
-	tenantID   string
-	tenantName string
-	authToken  string
-	roles      []*Role
-	catalog    []*Catalog
+	scopingType ScopingType
+	tenant      Tenant
+	domain      Domain
+	authToken   string
+	roles       []*Role
+	catalog     []*Catalog
+}
+
+//NewScopedToTenantAuthorization is a constructor for auth info scoped to tenant
+func NewScopedToTenantAuthorization(tenant Tenant, domain Domain, authToken string, roleIDs []string, catalog []*Catalog) Authorization {
+	roles := []*Role{}
+	for _, roleID := range roleIDs {
+		roles = append(roles, &Role{Name: roleID})
+	}
+	return &BaseAuthorization{
+		scopingType: ScopedToTenant,
+		tenant:      tenant,
+		domain:      domain,
+		roles:       roles,
+		authToken:   authToken,
+		catalog:     catalog,
+	}
 }
 
-//NewAuthorization is a constructor for auth info
-func NewAuthorization(tenantID, tenantName, authToken string, roleIDs []string, catalog []*Catalog) Authorization {
+//NewScopedToDomainAuthorization constructs auth info when user is scoped to domain
+func NewScopedToDomainAuthorization(domain Domain, authToken string, roleIDs []string, catalog []*Catalog) Authorization {
```

Comment:
code duplication, can we extract the common part? maybe with Functional Options, or a Builder?